### PR TITLE
Fixed warnings issued by new xt-clang compiler

### PR DESCRIPTION
**Pull request by aborisovich**
New Xtensa xt-clang driver issues new warnings that were not present before when using xt-xcc driver.
Fixed warnings related to enum cast to other enum. Silenced warnings related to C/C++ interoptability.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

### CODE REVIEW

**Review by marcinszkudlinski — APPROVED** [review]

File: src/audio/selector/selector.c [hunk 1]
```diff
@@ -579,7 +579,12 @@ static void build_config(struct comp_data *cd, struct module_config *cfg)
 	enum sof_ipc_frame __sparse_cache frame_fmt, valid_fmt;
 	int i;
 
-	frame_fmt = cfg->base_cfg.audio_fmt.depth;
+	/*
+	 * When compiling with CONFIG_IPC_MAJOR_4 depth is of
+	 * enum ipc4_bit_depth type and requires a cast to numeric type
+	 * before recasting to other enum.
+	 */
+	frame_fmt = (uint32_t)cfg->base_cfg.audio_fmt.depth;
```
Comment (paulstelian97) [inline src/audio/selector/selector.c:587]: Why do we need to have a conversion between different enums in the first place?

**Review by paulstelian97 — COMMENTED** [review]
Your change can hide a genuine bug (at least from a superficial analysis); see review comments.

I could probably approve this if given a good answer in that situation.

File: src/audio/selector/selector.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/selector/selector.c:587]: @aborisovich which enums are impacted here ? Sounds like we may need an alignment.

Comment (aborisovich) [reply src/audio/selector/selector.c:587]: Well, when `sof/lib/dai.h` is included, depth is a `uint32_t` type member of `dai_plat_fifo_data` struct.
However when we compile with IPC4 config option, header `src\include\ipc4\base-config.h` is uder and depth is `enum ipc4_bit_depth` type what causes the warning:

> implicit conversion from enumeration type 'enum ipc4_bit_depth' to different enumeration type 'enum sof_ipc_frame' [-Wenum-conversion]

Comment (aborisovich) [reply src/audio/selector/selector.c:587]: Yes, as @paulstelian97 indicated this looks like a serious bug.
Enums are implicitly cast to numerical values so what we do here,
is trying to assign one of the values from this set { 8, 16, 24, 32, 64, 65 } (values of enumerators of `enum ipc4_bit_depth`) as enumerator with such index to `enum sof_ipc_frame` which has enumerator values { 0, 1, 2, 3, 4 }.
I'm quite positive this is undefined behavior in C.

**Comment by aborisovich** [comment]
Converted back to draft, we need to verify & fix the case with enum casting.

**Comment by aborisovich** [comment]
Resolved issue with enum casting - converting back to normal PR.
**Resolution:** removed assignment from incompatible enumerator value:
`frame_fmt = cfg->base_cfg.audio_fmt.depth` &
`frame_fmt = cd->sink_format`
**Motivation:** assignment is not needed. Actually `frame_fmt` variable is not needed at all in the body of the:
`static void build_config(struct comp_data *cd, struct module_config *cfg)`
function. `audio_stream_fmt_conversion` function writes to `frame_fmt` as output, so this assignments are pointless.
Existence of `frame_fmt` is actually pointless too, as `cd->source_format` and corresponding `cd->sink_format` could be used
as arguments to `audio_stream_fmt_conversion` function.

However, this PR is not about refactoring but fixing the warnings issues by xt-clang.
Why warnings are issued by clang? Because they can.
Some explanation from C standard draft ISO/IEC 9899:TC3:

> Annex I
> (informative)
> Common warnings
> An implementation may generate warnings in many situations, none of which are
> specified as part of this International Standard. The following are a few of the more
> common situations.
> (...)
> - A value is given to an object of an enumerated type other than by assignment of an
> enumeration constant that is a member of that type, or an enumeration object that has
> the same type, or the value of a function that returns the same enumerated type
> (6.7.2.2).

Research done:
* debugged `build_ut/test/cmocka/src/audio/selector/selector_test` unit test with GDB.
Unfortunately, the test is not calling `static void build_config(...)` function that is a part of the `static int selector_init(...)` function, what is completely bizarre situation (unit tests not calling code under test?)
* debugged some end-to-end tests on MTL from "TestDmicMicselPeakvolHdaPipe" family hoping to resolve related bugs but fixing enum assignments did not bring any improvement.

File: src/audio/selector/selector.c [hunk 2]
```diff
@@ -579,14 +579,12 @@ static void build_config(struct comp_data *cd, struct module_config *cfg)
...
 	int i;
 
-	frame_fmt = cfg->base_cfg.audio_fmt.depth;
 	audio_stream_fmt_conversion(cfg->base_cfg.audio_fmt.depth,
 				    cfg->base_cfg.audio_fmt.valid_bit_depth,
 				    &frame_fmt, &valid_fmt,
 				    cfg->base_cfg.audio_fmt.s_type);
 	cd->source_format = frame_fmt;
 
-	frame_fmt = cd->sink_format;
```
Comment (marc-hb) [inline src/audio/selector/selector.c:589]: Should this file be in a separate commit?

Comment (aborisovich) [reply src/audio/selector/selector.c:589]: Why? It is related only to removing xt-clang warning.

Comment (marc-hb) [reply src/audio/selector/selector.c:589]: Even if it's not a functional change, it very much looks like one.

Comment (aborisovich) [reply src/audio/selector/selector.c:589]: Let's focus on what it actually is and not what it looks to be. As some people say "don't judge the book by the cover" :-)

**Review by pjdobrowolski — APPROVED** [review]

File: src/audio/selector/selector.c (hunk 2 above)
Comment (kv2019i) [reply src/audio/selector/selector.c:589]: @aborisovich You do need to explain  this in the git commit in some way why is it ok to remove these two lines. This is not unheard of to have code that doesn't do anything, but expectation is that you explain in commit to explain why this is the case.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Commit msg needs a bit more rationale...

**Review by abonislawski — APPROVED** [review]

File: src/audio/selector/selector.c (hunk 2 above)
Comment (aborisovich) [reply src/audio/selector/selector.c:589]: ok sure, I'll add some proper message

**Comment by aborisovich** [comment]
> build_config

Provided more description to commit message. Is this ok now @kv2019i ?

**Review by paulstelian97 — APPROVED** [review]
I'm fine with the commit, just a few minor things with the message (that I can allow).

First, "interoperability" is the right spelling.

The other thing is I would have appreciated some blank lines between the individual paragraphs on the commit description.

**Comment by aborisovich** [comment]
> First, "interoperability" is the right spelling.

Thanks for caching this, fixed the spelling. Added some line breaks too.

**Review by paulstelian97 — APPROVED** [review]
Re-approving, looks good now. I'd suggest for future work to eventually look at the enum issue you uncovered in a previous version of this.

**Review by softwarecki — APPROVED** [review]

**Comment by kv2019i** [comment]
This is racing with https://github.com/thesofproject/sof/pull/7058 but as this now has tests passing and reviews are ok, let's have this go in first. @aiChaoSONG you probably need to rebase.

One fail in https://sof-ci.01.org/sofpr/PR7096/build4004/devicetest/index.html in the suspend-resume case, but seems unrelated to this PR. Proceeding with merge.

**Review by kv2019i — APPROVED** [review]
